### ISSUE TITLE
EEPROM library: Does not work with WiFire boards in chipKIT-core 1.2.1

### ISSUE THREAD

**Issue by JacobChrist** [chipKIT32/chipKIT-core#254]
Tested with both with WiFire ECG and EFG silicon.

**Comment by JacobChrist** [comment]
Just thinking out loud here: The constant: _EEPROM_PAGE_COUNT in Deeprom.c is defined in System_Defs.h as such:

``` c++
#if defined(__PIC32MX1XX__) || defined(__PIC32MX2XX__)
	#define	_EEPROM_PAGE_SIZE	256		// In 32bit words
	#define	_EEPROM_PAGE_COUNT	4
#else
	#define	_EEPROM_PAGE_SIZE	1024	// In 32bit words
#define	_EEPROM_PAGE_COUNT	1
#endif
```
The following is from the PIC32MZ EF family manual:
Note: In PIC32MZ EF devices, the Flash page size is 16 KB (4K IW) and the row size is 2 KB (512 IW).

I don't know if this is causing the problem but it looks like a problem.

**Comment by JacobChrist** [comment]
Another possible issue, WiFire ADC uses ADC.  Though I'm not using the ADC in my simple example this comment was found in flash.c:

	// M00TODO: When DMA operations are supported in the core, need
	// to add code here to suspend DMA during the NVM operation.

**Comment by JacobChrist** [comment]
Another issue.  From Section 52 (Flash w/support for Live Updates) it indicates the unlock key is:
``` c++
 NVMKEY = 0x0;
 NVMKEY = 0xAA996655;
 NVMKEY = 0x556699AA;
```
The current unlock key is:
``` c++
 NVMKEY = 0xAA996655;
 NVMKEY = 0x556699AA;
```

Anyone know if we can just add the NVMKEY = 0x0; or should we wrap it in an ifdef for the MZ processors?

**Comment by JacobChrist** [comment]
Couple more notes:  In flash.c _doNvmOp there is a note about a needed 6us delay after writing to the NVMCON before unlocking memory.  I didn't see anything in either section 5 or 52 manuals about this being needed.  Additionally writing to this register is not mentioned in the section 52 manual so it may not be needed for the EC/EF parts and may actually be detrimental.  Another ifdef around this for MX parts?

I don't have time to test any of these findings tonight, dinner calls.

**Comment by JacobChrist** [comment]
In 52.6 ERROR CORRECTING CODE (ECC) AND FLASH PROGRAMMING it states that word programming is not allowed if FEC is enabled.  I can't imagine that it would be turned on by default but I haven't verified one way or the other.

**Comment by majenkotech** [comment]
NVMKEY = 0x0;

That is used to clear the key before unlocking. It's common to do it. It doesn't affect the actual unlocking, but it does ensure that it is actually locked before trying to unlock it, and that there is nothing odd in the key register before doing the unlocking that could affect the unlocking.

**Comment by majenkotech** [comment]
Ok, the problem we have here is actually two-fold.

First the EEPROM library needs to know the page size, which is simple enough:

````C
#if defined(__PIC32MX1XX__) || defined(__PIC32MX2XX__)
    #define _EEPROM_PAGE_SIZE   256     // In 32bit words
    #define _EEPROM_PAGE_COUNT  4
#elif defined(__PIC32MZ__)
    #define _EEPROM_PAGE_SIZE   4096
    #define _EEPROM_PAGE_COUNT  1
#else
    #define _EEPROM_PAGE_SIZE   1024    // In 32bit words
#define _EEPROM_PAGE_COUNT  1
#endif
````

Secondly the EEPROM flash memory is in the wrong place. At the moment it is located in KSEG0 which is not right. That is cached, which means that any writes to the EEPROM, while they actually work, won't be seen by the current sketch until either the board is reset or the cache block drops out the end of the cache and gets re-read.

So the EEPROM memory needs moving to KSEG1 to bypass the cache. That means changes not only to the core chipKIT-application-COMMON-MZ.ld but also to the linker scripts for the individual boards / chips. 

Basically all references to kseg0_eeprom_addr need changing to kseg1_eeprom_addr and the address changing from 0x9... to 0xB...

Yes, the "kseg0" bit is just a label, but naming it kseg0 and storing it in kseg1 seems a bit silly to me and can only cause confusion later on down the line.

**Comment by majenkotech** [comment]
My test program:

````C++
#include <EEPROM.h>

void setup() {
    Serial.begin(115200);
    Serial.printf("Current EEPROM 0 content:    0x%02X\r\n", EEPROM.read(0));
    uint8_t r = analogRead(A0);
    Serial.printf("Setting EEPROM 0 content to: 0x%02X\r\n", r);
    EEPROM.write(0, r);
    Serial.printf("Verify EEPROM 0 content:     0x%02X\r\n", EEPROM.read(0));
    Serial.println("Reset board to try again.");
}

void loop() {
}
````

And the output after making those changes:
````

Current EEPROM 0 content:    0x18
Setting EEPROM 0 content to: 0x6B
Verify EEPROM 0 content:     0x6B
Reset board to try again.
Current EEPROM 0 content:    0x6B
Setting EEPROM 0 content to: 0x5E
Verify EEPROM 0 content:     0x5E
Reset board to try again.
Current EEPROM 0 content:    0x5E
Setting EEPROM 0 content to: 0x38
Verify EEPROM 0 content:     0x38
Reset board to try again.
````

**Comment by majenkotech** [comment]
pr #256 created for you to test with.

**Comment by JacobChrist** [comment]
I tested pr #256 with a WiFire and it is working.  I suppose we should leave this open until its merged with the core.

**Comment by JacobChrist** [comment]
This mostly works in 1.3.1, there is a one-off error that is allowing writes to location 0x200.  The address writable should be 0x1FF.

```
0x000	00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F 
0x010	10 11 12 13 14 15 16 17 18 19 1A 1B 1C 1D 1E 1F 
0x020	20 21 22 23 24 25 26 27 28 29 2A 2B 2C 2D 2E 2F 
0x030	30 31 32 33 34 35 36 37 38 39 3A 3B 3C 3D 3E 3F 
0x040	40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F 
0x050	50 51 52 53 54 55 56 57 58 59 5A 5B 5C 5D 5E 5F 
0x060	60 61 62 63 64 65 66 67 68 69 6A 6B 6C 6D 6E 6F 
0x070	70 71 72 73 74 75 76 77 78 79 7A 7B 7C 7D 7E 7F 
0x080	80 81 82 83 84 85 86 87 88 89 8A 8B 8C 8D 8E 8F 
0x090	90 91 92 93 94 95 96 97 98 99 9A 9B 9C 9D 9E 9F 
0x0A0	A0 A1 A2 A3 A4 A5 A6 A7 A8 A9 AA AB AC AD AE AF 
0x0B0	B0 B1 B2 B3 B4 B5 B6 B7 B8 B9 BA BB BC BD BE BF 
0x0C0	C0 C1 C2 C3 C4 C5 C6 C7 C8 C9 CA CB CC CD CE CF 
0x0D0	D0 D1 D2 D3 D4 D5 D6 D7 D8 D9 DA DB DC DD DE DF 
0x0E0	E0 E1 E2 E3 E4 E5 E6 E7 E8 E9 EA EB EC ED EE EF 
0x0F0	F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 FA FB FC FD FE FF 
0x100	01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F 10 
0x110	11 12 13 14 15 16 17 18 19 1A 1B 1C 1D 1E 1F 20 
0x120	21 22 23 24 25 26 27 28 29 2A 2B 2C 2D 2E 2F 30 
0x130	31 32 33 34 35 36 37 38 39 3A 3B 3C 3D 3E 3F 40 
0x140	41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F 50 
0x150	51 52 53 54 55 56 57 58 59 5A 5B 5C 5D 5E 5F 60 
0x160	61 62 63 64 65 66 67 68 69 6A 6B 6C 6D 6E 6F 70 
0x170	71 72 73 74 75 76 77 78 79 7A 7B 7C 7D 7E 7F 80 
0x180	81 82 83 84 85 86 87 88 89 8A 8B 8C 8D 8E 8F 90 
0x190	91 92 93 94 95 96 97 98 99 9A 9B 9C 9D 9E 9F A0 
0x1A0	A1 A2 A3 A4 A5 A6 A7 A8 A9 AA AB AC AD AE AF B0 
0x1B0	B1 B2 B3 B4 B5 B6 B7 B8 B9 BA BB BC BD BE BF C0 
0x1C0	C1 C2 C3 C4 C5 C6 C7 C8 C9 CA CB CC CD CE CF D0 
0x1D0	D1 D2 D3 D4 D5 D6 D7 D8 D9 DA DB DC DD DE DF E0 
0x1E0	E1 E2 E3 E4 E5 E6 E7 E8 E9 EA EB EC ED EE EF F0 
0x1F0	F1 F2 F3 F4 F5 F6 F7 F8 F9 FA FB FC FD FE FF 00 
0x200	02 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x210	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x220	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x230	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x240	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x250	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x260	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x270	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x280	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x290	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x2F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x300	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x310	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x320	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x330	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x340	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x350	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x360	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x370	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x380	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x390	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x3F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x400	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x410	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x420	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x430	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x440	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x450	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x460	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x470	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x480	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x490	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x4F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x500	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x510	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x520	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x530	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x540	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x550	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x560	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x570	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x580	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x590	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x5F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x600	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x610	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x620	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x630	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x640	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x650	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x660	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x670	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x680	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x690	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x6F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x700	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x710	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x720	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x730	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x740	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x750	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x760	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x770	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x780	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x790	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x7F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x800	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x810	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x820	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x830	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x840	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x850	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x860	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x870	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x880	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x890	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x8F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x900	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x910	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x920	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x930	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x940	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x950	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x960	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x970	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x980	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x990	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9A0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9B0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9C0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9D0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9E0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0x9F0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xA90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAD0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xAF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xB90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBD0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xBF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xC90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCD0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xCF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xD90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDD0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xDF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xE90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xEA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xEB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xEC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xED0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xEE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xEF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF00	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF10	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF20	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF30	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF40	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF50	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF60	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF70	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF80	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xF90	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFA0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFB0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFC0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFD0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFE0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
0xFF0	FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 
```

The above output was created using this code:

```
#define MAX_ADDRESS_DEFAULT    4096

void eeprom_write(char *command)
{
  // there are only MAX_ADDRESS_DEFAULT bytes of EEPROM
  for(int address = 0; address < MAX_ADDRESS_DEFAULT; address++)
  {
    EEPROM.write(address, (byte)((address & 0xff) + ((address >> 8) & 0xff) ) );
  }
}

void eeprom_read(char *command)
{
  byte value;
  // there are only MAX_ADDRESS_DEFAULT bytes of EEPROM
  for(int address = 0; address < MAX_ADDRESS_DEFAULT; )
  {
    Serial.print("0x");
    if( address <= 0x0f ) Serial.print("0");
    if( address <= 0xff ) Serial.print("0");
    Serial.print(address, HEX);
    Serial.print("\t");
    for(int i = 0; i < 16; i++, address++)
    {
      // read a byte from the current address of the EEPROM
      value = EEPROM.read(address);
      if( value <= 0x0f ) Serial.print("0");
      Serial.print(value, HEX);
      Serial.print(" ");
    }
    Serial.println();
    delay(100);
  }
}
```

**Comment by JacobChrist** [comment]
Actually I'm going to close this one and open a separate issue for the one-off error.
